payu/envmod.py

```diff
@@ -109,4 +125,79 @@ def lib_update(required_libs, lib_name):
             return '{0}/{1}'.format(mod_name, mod_version)
 
     # If there are no libraries, return an empty string
-    return ''
\ No newline at end of file
+    return ''
+
+
+def setup_user_modules(user_modules, user_modulepaths):
+    """Run module use + load commands for user-defined modules"""
+
+    if 'MODULESHOME' not in os.environ:
+        print(
+            'payu: warning: No Environment Modules found; ' +
+            'skipping running module use/load commands for any module ' +
+            'directories/modulefiles defined in config.yaml')
+        return
+
+    # Add user-defined directories to MODULEPATH
+    for modulepath in user_modulepaths:
+        if not os.path.isdir(modulepath):
+            raise ValueError(
+                f"Module directory is not found: {modulepath}" +
+                "\n Check paths listed under `modules: use:` in config.yaml")
+
+        module('use', modulepath)
+
+    for modulefile in user_modules:
+        # Check module exists and there is not multiple available
+        module_subcommand = f"avail --terse {modulefile}"
+        output = run_cmd(module_cmd(module_subcommand)).stderr
+
+        # Extract out the modulefiles available
+        modules = [line for line in output.strip().splitlines()
+                   if not (line.startswith('/') and line.endswith(':'))]
+
+        if len(modules) > 1:
+            # Modules are used for finding model executable paths - so check
+            # for unique module -TODO: Could be a warning rather than an error?
+            raise ValueError(
+                f"There are multiple modules available for {modulefile}:\n" +
+                f"{output}\n{MULTIPLE_MODULES_HELP}")
+        elif len(modules) == 0:
+            raise ValueError(
+                f"Module is not found: {modulefile}\n{MODULE_NOT_FOUND_HELP}"
+            )
+
+        # Load module
+        module('load', modulefile)
+
+
+def env_var_set_by_modules(user_modules, env_var):
+    """Return an environment variable post loading only user-defined modules
+    - this is used for getting $PATH for searching for the model executable"""
+    if 'MODULESHOME' not in os.environ:
+        print('payu: warning: No Environment Modules found; skipping '
+              f'inspecting user module changes to ${env_var}')
+        return
+
+    # Note: Using subprocess shell to isolate changes to environment
+    load_commands = [f'load {module}' for module in user_modules]
+    commands = ['purge'] + load_commands
+    module_cmds = [f"eval `{module_cmd(c)}`" for c in commands]
+    module_cmds += [f'echo ${env_var}']
+    command = ' && '.join(module_cmds)
+    output = run_cmd(command)
```

Comment:
If this was a separate function it would be easily testable by checking the return string matched expectations.

That test might be quite uninformative though. I'll leave it up to you if you think it is worth it.

Reply:
On second thoughts, it would be great if we didn't have to do this at all.

Is there a reason we can't just save the `env` state before loading the user modules, and then comparing the `env` after we load the modules and see what has changed in `$PATH`?

Reply:
Yeah thats a good point. I had previously kept them separate as the user modules were loaded in `experiment.run`. But as they are being loaded in setup and `module purge` isn't needed, I changed it to inspect the `$PATH` and `$LOADEDMODULES` in the same function that loads the user modules. I've added in code that will `unload` the module is it was already loaded before `payu setup`, so it can pick up any changes to those environment variables